include/rms/rms.hh

```diff
@@ -1,4 +1,80 @@
+/* daft engineers vocoder project
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
 #ifndef RMS_RMS_HH
 #define RMS_RMS_HH
 
+#include "../audio/audio.hh"
+#include "../pipe.hh"
+#include <algorithm>
+#include <array>
+#include <cmath>
+#include <numeric>
+
+namespace rms {
+
+template <std::size_t num_samples> class RMS {
+  private:
+    Pipe<Audio> &input_pipe;
+    Pipe<double> &output_pipe;
+
+    std::array<uint32_t, num_samples> sample_buffer = {0};
+    int sample_buffer_index = 0;
+
+    bool run_thread{};
+    std::thread thread;
+
+  public:
+    RMS(Pipe<Audio> &input_pipe, Pipe<double> &output_pipe) : input_pipe(input_pipe), output_pipe(output_pipe) {
+    }
+
+    void insert(Audio packet) {
+        std::for_each(packet.begin(), packet.end(), [this](const uint16_t item) {
+            // this assumes that the 0 point for the signal is max(uint16)/2, precicely in the middle of the range
+
+            int64_t centred = item - UINT16_MAX / 2;
```

Comment:
Might as well choose a signed 32/16 bit number here so that we aren't 64 bit dependant? No reason this couldn't run on an older Pi

Reply:
Actually, even better: we should push forward with changing Audio to be a signed 16 bit value, which should simplify this a bit anyway.